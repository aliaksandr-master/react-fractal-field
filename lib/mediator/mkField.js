export const mkMeta = () => ({
  valid: true,
  touched: false,

  active: false,
  activated: false,

  submitted: false,
  submitting: false,
  submitFailed: false,
  submitSuccess: false,
  submittedTimes: 0,

  error: null,
  errorsIn: [],
  hasException: false,
  hasFormatException: false,
  hasNormalizeException: false,
  hasValidateException: false
});

export default ({
  id,
  complexValueType = null,
  parentID = null,
  autoClean = false,
  shareMeta = true,
  shareValue = true,
  value = undefined,
  name = null
}) => ({
  // permanent
  id,
  parentID,
  autoClean,
  shareMeta,
  shareValue,
  complexValueType, // if not null component might have named children

  // free transform
  name,

  value,

  meta: mkMeta(),

  children: [],
  version: 0,
  metaVersion: 0,
  valueVersion: 0
});
