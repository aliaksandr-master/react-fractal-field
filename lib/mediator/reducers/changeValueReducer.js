import { isEqual } from '../../utils/lodash';
import recalcChildrenValue from '../recalcChildrenValue';
import recalcValue from '../recalcValue';
import { updateFieldValue, fieldReducer } from '../state-helpers';



export default () => fieldReducer((state, field, value) => {
  if (isEqual(field.value, value)) {
    return state;
  }

  state = updateFieldValue(state, field, value);

  state = recalcChildrenValue(state, field.id);

  state = recalcValue(state, field.parentID, field.id);

  return state;
});
