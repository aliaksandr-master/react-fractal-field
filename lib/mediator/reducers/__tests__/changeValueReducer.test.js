/*eslint-env jest*/
import changeValueReducer from '../changeValueReducer';
import mkField from '../../mkField';


test('changeValueReducer', () => {
  expect(changeValueReducer()({}, { type: 'some', fieldID: 123, payload: 'some' })).toMatchSnapshot();

  expect(changeValueReducer()({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123, payload: 'some' })).toMatchSnapshot();

  expect(changeValueReducer()({ [123]: mkField({ id: 123, value: 'some' }) }, { type: 'some', fieldID: 123, payload: 'some' })).toMatchSnapshot();
});
