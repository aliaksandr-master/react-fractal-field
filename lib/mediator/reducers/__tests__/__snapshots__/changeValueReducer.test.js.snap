// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changeValueReducer 1`] = `Object {}`;

exports[`changeValueReducer 2`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": false,
      "hasFormatException": false,
      "hasNormalizeException": false,
      "hasValidateException": false,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": true,
    },
    "metaVersion": 0,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": "some",
    "valueVersion": 1,
    "version": 1,
  },
}
`;

exports[`changeValueReducer 3`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": false,
      "hasFormatException": false,
      "hasNormalizeException": false,
      "hasValidateException": false,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": true,
    },
    "metaVersion": 0,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": "some",
    "valueVersion": 0,
    "version": 0,
  },
}
`;
